download.go

```diff
@@ -0,0 +1,244 @@
+		Prefix:    path,
+		Recursive: recursive,
+	})
+	result := make([]string, 0, len(objectCh))
+
+	for obj := range objectCh {
+		if obj.Err != nil {
+			return nil, obj.Err
+		}
+
+		fileName := strings.TrimPrefix(obj.Key, path)
+		result = append(result, fileName)
+	}
+
+	return result, nil
+}
+
+func (c *client) DownloadFile(ctx context.Context, path string) (*File, error) {
+	return c.DownloadFileWithOptions(ctx, path, minio.GetObjectOptions{})
+}
+
+func (c *client) DownloadFileWithOptions(ctx context.Context, path string, options minio.GetObjectOptions) (*File, error) {
+	const errMessage = "failed to download file from s3: %w"
+
+	object, err := c.minioClient.GetObject(ctx, c.bucketName, path, options)
+	if err != nil {
+		return nil, fmt.Errorf(errMessage, err)
+	}
+
+	defer object.Close()
+
+	fileInfo, err := object.Stat()
+	if err != nil {
+		return nil, fmt.Errorf(errMessage, err)
+	}
+
+	content := make([]byte, fileInfo.Size)
+
+	_, err = object.Read(content)
+	if err != nil && !errors.Is(err, io.EOF) {
+		var minioResponse minio.ErrorResponse
+
+		if errors.As(err, &minioResponse) {
+			return nil, ErrNotFound
+		}
+
+		return nil, fmt.Errorf(errMessage, err)
+	}
+
+	return &File{
+		Content:      content,
+		ModifiedDate: fileInfo.LastModified,
+		ContentType:  fileInfo.ContentType,
+		Name:         pathpkg.Base(path),
+		MetaData:     fileInfo.UserMetadata,
+		Checksum:     fileInfo.ChecksumCRC32C,
+		Length:       fileInfo.Size,
+	}, nil
+}
+
+func (c *client) DownloadFileToLocalPath(ctx context.Context, path, localPath string) error {
+	return c.DownloadFileToLocalPathWithOptions(ctx, path, localPath, minio.GetObjectOptions{})
+}
+
+func (c *client) DownloadFileToLocalPathWithOptions(ctx context.Context, path, localPath string, options minio.GetObjectOptions) error {
+	return c.minioClient.FGetObject( //nolint:wrapcheck
+		ctx,
+		c.bucketName,
+		path,
+		localPath,
+		options,
+	)
+}
+
+func (c *client) DownloadDirectory(ctx context.Context, path string) ([]*File, error) {
+	return c.DownloadDirectoryWithOptions(ctx, path, minio.GetObjectOptions{})
+}
+
+func (c *client) DownloadDirectoryWithOptions(ctx context.Context, path string, options minio.GetObjectOptions) ([]*File, error) {
+	const errMessage = "failed to download directory: %w"
+
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+
+	objectCh := c.minioClient.ListObjects(ctx, c.bucketName, minio.ListObjectsOptions{
+		Prefix:    path,
+		Recursive: true,
+	})
+
+	wg := new(sync.WaitGroup)
+	errCh := make(chan error)
+	result := make([]*File, 0, len(objectCh))
+
+	for obj := range objectCh {
+		if obj.Err != nil {
+			return nil, fmt.Errorf(errMessage, obj.Err)
+		}
+
+		wg.Add(1)
+
+		go func(obj minio.ObjectInfo) {
+			doc, err := c.DownloadFileWithOptions(ctx, obj.Key, options)
+			if err != nil {
+				errCh <- err
+
+				return
+			}
+
+			result = append(result, doc)
+
+			wg.Done()
```

Comment:
falls etwas über dem `wg.Done()` schief läuft rufen wir es niemals auf.. 
die waitgroup wächst dann immer weiter :see_no_evil: 

```suggestion
			defer wg.Done()
			
			doc, err := c.DownloadFileWithOptions(ctx, obj.Key, options)
			if err != nil {
				errCh <- err

				return
			}

			result = append(result, doc)
```